datafusion/core/src/physical_plan/union.rs

```diff
@@ -78,10 +83,24 @@ impl UnionExec {
             inputs[0].schema().metadata().clone(),
         ));
 
+        // If all the input partitions have the same Hash partition spec with the first_input_partition
```

Comment:
I may have missed it, but I don't see any tests for this code (either new tests or changes in existing tests).